### PR TITLE
feat: simple join

### LINKED ISSUE #175: Support simple join for associate Table.
**What would you like to be added**:

**Why is this needed**:

Support simple join for associate Table.

I think we should support simple join, It only can join between tables for the same sharding algorithm at same physical database instance.

### PR DESCRIPTION
**Pull request by ZLBer**
**What this PR does**:
 support A join B in one db
shard table join shard table
shard table join no shard table

**Which issue(s) this PR fixes**:

Fixes #175

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note

```

### CODE REVIEW

**Review by AlexStocks — APPROVED** [review]

File: pkg/runtime/optimize/optimizer.go [hunk 1]
```diff
@@ -322,6 +324,76 @@ func (o optimizer) optimizeSelect(ctx context.Context, conn proto.VConn, stmt *r
...
+		return nil, errors.WithStack(errNoRuleFound)
+	}
+
+	join := stmt.From[0].Source().(*rast.JoinNode)
+
+	compute := func(tableSource *rast.TableSourceNode) (alias string, shardList []string, err error) {
+		table := tableSource.TableName()
+		alias = tableSource.Alias()
+		if table == nil {
+			err = errors.New("must table，not statement or join node")
```
Comment (jjeffcaii) [inline pkg/runtime/optimize/optimizer.go:341]: Use EN comma instead.

File: pkg/runtime/optimize/optimizer.go [hunk 2]
```diff
@@ -322,6 +324,76 @@ func (o optimizer) optimizeSelect(ctx context.Context, conn proto.VConn, stmt *r
...
+
+	aliasLeft, shardLeft, err := compute(join.Left)
+	if err != nil {
+		return nil, err
+	}
+	aliasRight, shardRight, err := compute(join.Right)
+
+	if err != nil {
+		return nil, err
+	}
```
Comment (jjeffcaii) [inline pkg/runtime/optimize/optimizer.go:378]: Should check left/right shards are in the same db group.

**Review by jjeffcaii — APPROVED** [review]

File: pkg/runtime/optimize/optimizer.go [hunk 3]
```diff
@@ -322,6 +324,83 @@ func (o optimizer) optimizeSelect(ctx context.Context, conn proto.VConn, stmt *r
...
+	compute := func(tableSource *rast.TableSourceNode) (database, alias string, shardList []string, err error) {
+		table := tableSource.TableName()
+		alias = tableSource.Alias()
+		database = table.Prefix()
+
+		if table == nil {
+			err = errors.New("must table, not statement or join node")
+			return
+		}
+		shards, err := o.computeShards(ru, table, nil, args)
```
Comment (baerwang) [inline pkg/runtime/optimize/optimizer.go:346]:
```go
table := tableSource.TableName()

if table == nil {
	err = errors.New("must table, not statement or join node")
	return
}

alias = tableSource.Alias()
database = table.Prefix()
```

Comment (ZLBer) [reply pkg/runtime/optimize/optimizer.go:346]: fix

**Review by baerwang — APPROVED** [review]

**Review by dongzl — APPROVED** [review]

**Pull request merged**